test/e2e/example/example.go

```diff
@@ -0,0 +1,133 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package example
+
+import (
+    "fmt"
+    "time"
+
+    "github.com/rancher/submariner/test/e2e/framework"
+
+    v1 "k8s.io/api/core/v1"
+    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+    "k8s.io/apimachinery/pkg/util/wait"
+
+    . "github.com/onsi/ginkgo"
+    . "github.com/onsi/gomega"
+)
+
+var _ = Describe("[example] Basic example to demonstrate how to write tests using the framework", func() {
+    f := framework.NewDefaultFramework("basic-example")
+    It("Should be able to list existing nodes on the cluster", func() {
+        testListingNodes(f)
+    })
+    It("Should be able to create a pod using the provided client", func() {
+        testCreatingAPod(f)
+    })
+})
+
+func testListingNodes(f *framework.Framework) {
+    cs := f.ClientSet
+    nc := cs.CoreV1().Nodes()
+    By("Requesting node list from API")
+    nodes, err := nc.List(metav1.ListOptions{})
+    Expect(err).NotTo(HaveOccurred())
+
+    By("Checking that we had more than 0 nodes on the reponse")
+    Expect(len(nodes.Items)).ToNot(BeZero())
+
+    for _, node := range nodes.Items {
+        inIP, err := getIP(v1.NodeInternalIP, &node)
+        Expect(err).NotTo(HaveOccurred())
+        framework.Logf("Detected node with IP: %v", inIP)
+    }
+}
+
+var (
+    testPod = v1.Pod{
+        ObjectMeta: metav1.ObjectMeta{
+            GenerateName: "example-pod",
+            Labels: map[string]string{
+                "example-pod": "",
+            },
+        },
+        Spec: v1.PodSpec{
+            Containers: []v1.Container{
+                {
+                    Name:  "example-pod",
+                    Image: "busybox",
+                    Command:  []string{"sh", "-c", "echo Hello Kubernetes, I am at $POD_IP! && sleep 3600"},
+                    Env: []v1.EnvVar{
+                        {
+                            Name:      "POD_IP",
+                            ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "status.podIP"}},
+                        },
+                    },
+                },
+            },
+        },
+    }
+)
+
+func testCreatingAPod(f *framework.Framework) {
+    cs := f.ClientSet
+    pc := cs.CoreV1().Pods(f.Namespace.Name)
+
+    By("Creating a bunch of pods")
+    for i := 0; i < 3; i ++ {
+        _, err := pc.Create(&testPod)
+        Expect(err).NotTo(HaveOccurred())
+    }
+
+    By("Waiting for the example-pod(s) to be scheduled and running")
+    err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (bool, error) {
+        pods, err := pc.List(metav1.ListOptions{LabelSelector: "example-pod"})
+        if err != nil {
+            framework.Logf("Transient failure when attempting to list pods: %v", err)
```

Comment:
(No action required) Apologies for not being clear - that the error may well represent a permanent failure and failing fast may be desirable. Inspection with some of the methods in `k8s.io/apimachinery/pkg/api/errors` is suggested.  My comment was more that it may be desirable to retry on transient failures instead of exiting the polling loop.

Reply:
Thanks, no problem, looking at that package would it make sense to consider "http.StatusInternalServerError" ones as transient?

Reply:
I don't see a clear function to identify transient errors, I could combine (or):
- IsInternalError
- IsServerTimeout
- IsTimeout
- IsServiceUnavailable
- IsUnexpectedServerError
- IsTooManyRequests

into a helper function for the tests may be.

Reply:
Sure, such a helper could be useful.  Whether the error is considered recoverable or not might depend on the circumstance.  I'm not trying to be prescriptive, though - just trying to raise awareness of the implications of returning / not returning errors in a wait loop.

Reply:
Of course, thanks, I think there's value in making testing as resistant to transient errors as possible. While trying to keep it as clean as possible

Reply:
Ok, I didn't resolve this bit yet. I want to do it now or on follow up, because I believe it's valuable.

Reply:
Ok, resolved now.